Source/DafnyCore/AST/PreprocessorHelper.cs

```diff
@@ -0,0 +1,196 @@
+        if (ignoreCutoff != -1) {
+          // we're already in a state of ignoring, so continue to ignore
+        } else if (IfdefConditionSaysToInclude(t.Substring(3).TrimStart(), defines)) {
+          // include this branch
+        } else {
+          ignoreCutoff = readState.Count; // start ignoring
+          rs.mayStillIncludeAnotherAlternative = true; // allow some later "elsif" or "else" branch to be included
+        }
+
+        readState.Add(rs);
+        sb.Append(newline); // ignore the #if line
+      } else if (t.StartsWith("#elsif")) {
+        ReadState rs;
+        if (readState.Count == 0 || (rs = readState[readState.Count - 1]).hasSeenElse) {
+          sb.Append("#MalformedInput: misplaced #elsif" + newline); // malformed input
+          break;
+        }
+
+        if (ignoreCutoff == -1) {
+          // we had included the previous branch
+          //Contract.Assert(!rs.mayStillIncludeAnotherAlternative);
+          ignoreCutoff = readState.Count - 1; // start ignoring
+        } else if (rs.mayStillIncludeAnotherAlternative &&
+                   IfdefConditionSaysToInclude(t.Substring(6).TrimStart(), defines)) {
+          // include this branch, but no subsequent branch at this level
+          ignoreCutoff = -1;
+          rs.mayStillIncludeAnotherAlternative = false;
+          readState[readState.Count - 1] = rs;
+        }
+
+        sb.Append(newline); // ignore the #elsif line
+      } else if (t == "#else") {
+        ReadState rs;
+        if (readState.Count == 0 || (rs = readState[readState.Count - 1]).hasSeenElse) {
+          sb.Append("#MalformedInput: misplaced #else" + newline); // malformed input
+          break;
+        }
+
+        rs.hasSeenElse = true;
+        if (ignoreCutoff == -1) {
+          // we had included the previous branch
+          //Contract.Assert(!rs.mayStillIncludeAnotherAlternative);
+          ignoreCutoff = readState.Count - 1; // start ignoring
+        } else if (rs.mayStillIncludeAnotherAlternative) {
+          // include this branch
+          ignoreCutoff = -1;
+          rs.mayStillIncludeAnotherAlternative = false;
+        }
+
+        readState[readState.Count - 1] = rs;
+        sb.Append(newline); // ignore the #else line
+      } else if (t == "#endif") {
+        if (readState.Count == 0) {
+          sb.Append("#MalformedInput: misplaced #endif" + newline); // malformed input
+          break;
+        }
+
+        readState.RemoveAt(readState.Count - 1); // pop
+        if (ignoreCutoff == readState.Count) {
+          // we had ignored the branch that ends here; so, now we start including again
+          ignoreCutoff = -1;
+        }
+
+        sb.Append(newline); // ignore the #endif line
+      } else if (ignoreCutoff == -1) {
+        sb.Append(s);
+        sb.Append(newline);
+      } else {
+        sb.Append(newline); // ignore the line
+      }
+    }
+
+    return sb.ToString();
+  }
+
+  public static string DetectNewline(string filename) {
```

Comment:
Renamed to `GetNewlinesStyle` and added a comment.
I don't think I want to change the filename to `TextReader` because
* the `TextReader` resource I'm creating (a `StreamReader`) can be disposed at the end of the function, it's an implementation detail (otherwise, the caller would have to take care of this disposal),
* I'm using `reader.EndOfStream` which is not available on `TextReader`
* I would advocate to make things reusable on a case-by-case basis for internal functions such as this one. I am not marking it as internal because other projects might want to depend on it, for example plugins.